src/dataflow/operators/ros/to_ros_operator.rs

```diff
@@ -0,0 +1,76 @@
+use crate::dataflow::{context::SinkContext, operator::Sink, Data};
+use serde::Deserialize;
+use std::sync::Arc;
+
+/// Takes an input ERDOS stream and publishes to a ROS topic using the provided message conversion
+/// function.
+///
+/// Conversion function should convert a Rust data type and return a ROS type which implements
+/// the [`rosrust::Message`] trait.
+///
+/// # Example
+/// The following example shows how to use a [`ToRosOperator`] with a conversion function which takes
+/// a Rust [`i32`] and converts it to a ROS message with [`rosrust_msg::std_msgs::Int32`] data.
+///
+/// Assume that `source_stream` is an ERDOS stream sending the correct messages.
+///
+/// ```
+/// fn erdos_int_to_ros_int(input: &i32) -> rosrust_msg::std_msgs::Int32 {
+///     rosrust_msg::std_msgs::Int32 { data: input.data }
+/// }
+///
+/// let ros_sink_config = OperatorConfig::new().name("ToRosInt32");
+/// erdos::connect_sink(
+///     move || -> ToRosOperator<i32, rosrust_msg::std_msgs::Int32> {
+///         ToRosOperator::new("int_topic", erdos_int_to_ros_int)
+///     },
+///     || {},
+///     ros_sink_config,
+///     &source_stream,
+/// );
+/// ```
+
+pub struct ToRosOperator<T, U: rosrust::Message>
+where
+    T: Data + for<'a> Deserialize<'a>,
+{
+    publisher: rosrust::Publisher<U>,
+    to_ros_msg: Arc<dyn Fn(&T) -> U + Send + Sync>,
+}
+
+impl<T, U: rosrust::Message> ToRosOperator<T, U>
+where
+    T: Data + for<'a> Deserialize<'a>,
+{
+    pub fn new<F>(topic: &str, to_ros_msg: F) -> Self
+    where
+        F: 'static + Fn(&T) -> U + Send + Sync,
+    {
+        rosrust::init("publisher");
+        Self {
+            publisher: rosrust::publish(topic, 2).unwrap(),
+            to_ros_msg: Arc::new(to_ros_msg),
+        }
+    }
+}
+
+impl<T, U: rosrust::Message> Sink<(), T> for ToRosOperator<T, U>
+where
+    T: Data + for<'a> Deserialize<'a>,
+{
+    fn on_data(&mut self, ctx: &mut SinkContext<()>, data: &T) {
+        let timestamp = ctx.get_timestamp().clone();
+        let msg = (self.to_ros_msg)(data);
+
+        tracing::info!(
+            "{} @ {:?}: Sending {:?}",
+            ctx.get_operator_config().get_name(),
+            timestamp,
+            msg,
+        );
+        // Publishes converted message on topic.
+        self.publisher.send(msg).unwrap();
```

Comment:
This might be a topic for future discussion, but I'm wondering if we should check if there are any subscribers on the topic that we're publishing? ERDOS aims to ensure a dataflow execution that is free of any message loss. However, this guarantee breaks down once we touch the ROS interface.

We may also be able to require the number of subscribers that an ERDOS operator expects to have to ensure that there is no message loss.

Caveat: Even with these changes, we still cannot guarantee a loss-free communication since ROS can choose to arbitrarily drop messages from its queue at the Subscriber end.

Reply:
The rosrust library we're using allows for this -- See this [example](https://github.com/adnanademovic/rosrust/blob/3c5edfaa0d5c2f7528a50966f2ff6c1bb1bf7ea1/rosrust/examples/publisher.rs#L35).